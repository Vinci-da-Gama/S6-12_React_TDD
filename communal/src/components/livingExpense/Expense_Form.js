import React, { Component } from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import { Button, Alert } from 'reactstrap';

class ExpenseForm extends Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        };
    }

    onDescriptionChange(e) {
        const description = e.target.value;
        this.setState(() => ({ description }));
    }

    onAmountChange(e) {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    }

    onDateChange(createdAt) {
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    }

    onFocusChange({ focused }) {
        this.setState(() => ({ calendarFocused: focused }));
    }

    onNoteChange(e) {
        const note = e.target.value;
        this.setState(() => ({ note }));
    }

    onFormSubmit(e) {
        e.preventDefault();
        const error = 'Please provide description and amount.';
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error }));
        } else {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    }
    
    render () {
        return (
            <div>
                { this.state.error && <Alert color="danger">{this.state.error}</Alert> }
                <form noValidate name="expenseForm" onSubmit={ (event) => { this.onFormSubmit(event) } }>
                    <input type="text" name="ef_description" placeholder="Description..." autoFocus
                        value={this.state.description} onChange={ (event) => { this.onDescriptionChange(event) }} />
                    <input type="text" name="ef_Amount" placeholder="Amount..."
                        value={this.state.amount} onChange={ (event) => { this.onAmountChange(event) } } />
                    <SingleDatePicker date={this.state.createdAt} focused={this.state.calendarFocused}
                        onDateChange={ this.onDateChange } onFocusChange={ this.onFocusChange }
                        numberOfMonths={1} isOutsideRange={ () => false } />
                    <textarea name="ef_note" rows="5"
                        placeholder="Add a note for your expense (optional)"
                        value={this.state.note}
                        onChange={ (event) => this.onNoteChange(event) }></textarea>
                    <Button color="success" outline block>
                        Add_Expense
                    </Button>
                </form>
            </div>
        );
    }
};

export default ExpenseForm;