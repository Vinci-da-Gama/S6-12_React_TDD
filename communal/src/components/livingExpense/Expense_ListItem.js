import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';
import { ListGroup, ListGroupItem, Badge, Label } from 'reactstrap';

const ExpenseListItem = ({ id, description, amount, createdAt }) => {
    return (
        <div>
            <ListGroup>
                <ListGroupItem className="justify-content-between">
                    <Link to={`/edit/${id}`}>
                        <h3>{description}</h3>
                    </Link>
                    <i>Date: </i>{moment(createdAt).format('MMMM Do, YYYY')}
                    <Badge pill>
                        <Label>Amount: </Label>
                        {numeral(amount / 100).format('$0,0.00')}
                    </Badge>
                </ListGroupItem>
            </ListGroup>
        </div>
    );
};

export default ExpenseListItem;
