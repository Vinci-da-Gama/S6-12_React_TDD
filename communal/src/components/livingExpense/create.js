import React, { Component } from 'react';

class CreateExpense extends Component {
    render () {
        return (
            <div>
                create exp....
            </div>
        );
    }
};

export default CreateExpense;