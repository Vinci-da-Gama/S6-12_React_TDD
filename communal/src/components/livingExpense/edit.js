import React, { Component } from 'react';

class EditExpense extends Component {
    render () {
        return (
            <div>
                edit exp....
            </div>
        );
    }
};

export default EditExpense;