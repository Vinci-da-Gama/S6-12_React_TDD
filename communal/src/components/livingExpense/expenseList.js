import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Button, Alert } from 'reactstrap';

import selectExpenses from '../../helpers/data-handler/expense-filterNsort';
import ExpenseListItem from './Expense_ListItem';

export class ExpenseListCompo extends Component {
    render() {
        if (this.props.expenses.length === 0 || !this.props.expenses) {
            return (
                <Alert color="danger">
                    No expense item list.
                </Alert>
            );
        }
        return (
            <div>
                {
                    this.props.expenses.map((expense, idx) => {
                        return <ExpenseListItem key={ expense.id+idx } {...expense} />
                    })
                }
            </div>
        );
    }
}

const mapStateToProps = (state) => ({
    expenses: selectExpenses(state.expenses, state.setValViaFilters)
});

/* const mapDispatchToProps = { 
} */

export default connect(mapStateToProps, null)(ExpenseListCompo);
