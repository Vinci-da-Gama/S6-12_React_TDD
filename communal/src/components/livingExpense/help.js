import React, { Component } from 'react';

class HelpExpense extends Component {
    render () {
        return (
            <div>
                help exp....
            </div>
        );
    }
};

export default HelpExpense;