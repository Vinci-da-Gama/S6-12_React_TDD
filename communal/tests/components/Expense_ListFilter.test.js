import React from 'react';
import { shallow } from 'enzyme';
import moment from 'moment';

import { ExpenseListFilter } from '../../src/components/livingExpense/Expense_ListFilter';
import { defaultFiltersConditions, filtersWithData } from '../fixtures/expense-filters.fixture';

let setTextFilterSpy, sortByDateSpy, sortByAmountSpy, setStartDateSpy, setEndDateSpy, wrapper;

beforeEach(() => {
    setTextFilterSpy = jest.fn();
    sortByDateSpy = jest.fn();
    sortByAmountSpy = jest.fn();
    setStartDateSpy = jest.fn();
    setEndDateSpy = jest.fn();
    wrapper = shallow(<ExpenseListFilter 
        setValViaFilters={defaultFiltersConditions}
        setTextFilter = {setTextFilterSpy}
        sortByDate = {sortByDateSpy}
        sortByAmount = {sortByAmountSpy}
        setStartDate = {setStartDateSpy}
        setEndDate = {setEndDateSpy}
    />);
});

describe('Test snapshot with default filters condition and filters has data.', () => {
    test('should render expense list filter correctly.', () => {
        expect(wrapper).toMatchSnapshot();
    });
    test('should render expense list filters with data correctly.', () => {
        wrapper.setProps({
            setValViaFilters: filtersWithData
        });
        expect(wrapper).toMatchSnapshot();
    });
});

test('should handle set text filter.', () => {
    const value = 'rent';
    wrapper.find('#lf_textfilterId').simulate('change', {
        target: { value }
    });
    expect(setTextFilterSpy).toHaveBeenLastCalledWith(value);
});

test('should filter by sort by date.', () => {
    const value = 'date';
    wrapper.find('#lf_sortbydateOramountId').simulate('change', {
        target: { value }
    });
    expect(sortByDateSpy).toHaveBeenCalled();
});

test('should filter by sort by amount.', () => {
    const value = 'amount';
    wrapper.find('#lf_sortbydateOramountId').simulate('change', {
        target: { value }
    });
    expect(sortByAmountSpy).toHaveBeenCalled();
});

test('should handle date changed.', () => {
    const startDate = moment(0).add(4, 'years');
    const endDate = moment(0).add(8, 'years');
    wrapper.find('[startDateId="DpStartDate"]').prop('onDatesChange')({ startDate, endDate });
    expect(setStartDateSpy).toHaveBeenLastCalledWith(startDate);
    expect(setEndDateSpy).toHaveBeenLastCalledWith(endDate);
});

test('should handle focus changed.', () => {
    const calendarFocused = 'endDate';
    wrapper.find('[startDateId="DpStartDate"]').prop('onFocusChange')(calendarFocused);
    expect(wrapper.state('calendarFocused')).toBe(calendarFocused);
});
