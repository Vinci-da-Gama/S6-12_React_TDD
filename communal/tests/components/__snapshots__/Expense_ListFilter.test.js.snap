// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test snapshot with default filters condition and filters has data. should render expense list filter correctly. 1`] = `
<div
  className="row mx-3"
>
  <Col
    md={4}
    sm={6}
    tag="div"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    <FormGroup
      tag="div"
    >
      <Label
        for="lf_textfilterId"
        tag="label"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        Text_Filter
      </Label>
      <Input
        className="rounded-0"
        id="lf_textfilterId"
        name="lf_textfilter"
        onChange={[Function]}
        placeholder="Input Text For filtering..."
        type="text"
        value=""
      />
    </FormGroup>
  </Col>
  <Col
    md={4}
    sm={6}
    tag="div"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    <FormGroup
      tag="div"
    >
      <Label
        for="exampleSelectMulti"
        tag="label"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        Sorted_By
      </Label>
      <Input
        className="rounded-0 select-rightcorner"
        id="lf_sortbydateOramountId"
        name="lf_sortbydateOramount"
        onChange={[Function]}
        type="select"
        value="date"
      >
        <option
          value="date"
        >
          Date
        </option>
        <option
          value="amount"
        >
          Amount
        </option>
      </Input>
    </FormGroup>
  </Col>
  <Col
    md={4}
    sm={6}
    tag="div"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    <Label
      tag="label"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      StartDate -- EndDate
    </Label>
    <withStyles(DateRangePicker)
      anchorDirection="left"
      appendToBody={false}
      block={false}
      calendarInfoPosition="bottom"
      customArrowIcon={null}
      customCloseIcon={null}
      customInputIcon={null}
      daySize={39}
      disableScroll={false}
      disabled={false}
      displayFormat={[Function]}
      enableOutsideDays={false}
      endDate={null}
      endDateId="DpEndDate"
      endDatePlaceholderText="End Date"
      firstDayOfWeek={null}
      focusedInput={null}
      hideKeyboardShortcutsPanel={false}
      horizontalMargin={0}
      initialVisibleMonth={null}
      inputIconPosition="before"
      isDayBlocked={[Function]}
      isDayHighlighted={[Function]}
      isOutsideRange={[Function]}
      isRTL={false}
      keepFocusOnInput={false}
      keepOpenOnDateSelect={false}
      minimumNights={1}
      monthFormat="MMMM YYYY"
      navNext={null}
      navPrev={null}
      noBorder={false}
      numberOfMonths={1}
      onClose={[Function]}
      onDatesChange={[Function]}
      onFocusChange={[Function]}
      onNextMonthClick={[Function]}
      onPrevMonthClick={[Function]}
      openDirection="down"
      orientation="horizontal"
      phrases={
        Object {
          "calendarLabel": "Calendar",
          "chooseAvailableEndDate": [Function],
          "chooseAvailableStartDate": [Function],
          "clearDates": "Clear Dates",
          "closeDatePicker": "Close",
          "dateIsSelected": [Function],
          "dateIsUnavailable": [Function],
          "enterKey": "Enter key",
          "escape": "Escape key",
          "focusStartDate": "Interact with the calendar and add the check-in date for your trip.",
          "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
          "homeEnd": "Home and end keys",
          "jumpToNextMonth": "Move forward to switch to the next month.",
          "jumpToPrevMonth": "Move backward to switch to the previous month.",
          "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
          "keyboardShortcuts": "Keyboard Shortcuts",
          "leftArrowRightArrow": "Right and left arrow keys",
          "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
          "moveFocusByOneMonth": "Switch months.",
          "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
          "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
          "openThisPanel": "Open this panel.",
          "pageUpPageDown": "page up and page down keys",
          "questionMark": "Question mark",
          "returnFocusToInput": "Return to the date input field.",
          "selectFocusedDate": "Select the date in focus.",
          "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
          "upArrowDownArrow": "up and down arrow keys",
        }
      }
      readOnly={false}
      regular={false}
      renderCalendarInfo={null}
      renderDayContents={null}
      renderMonthElement={null}
      renderMonthText={null}
      reopenPickerOnClearDates={false}
      required={false}
      screenReaderInputMessage=""
      showClearDates={true}
      showDefaultInputIcon={false}
      small={false}
      startDate={null}
      startDateId="DpStartDate"
      startDatePlaceholderText="Start Date"
      verticalHeight={null}
      verticalSpacing={22}
      weekDayFormat="dd"
      withFullScreenPortal={false}
      withPortal={false}
    />
    <FormText
      color="secondary"
      tag="small"
    >
      Select Date Range for expense.
    </FormText>
  </Col>
</div>
`;

exports[`Test snapshot with default filters condition and filters has data. should render expense list filters with data correctly. 1`] = `
<div
  className="row mx-3"
>
  <Col
    md={4}
    sm={6}
    tag="div"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    <FormGroup
      tag="div"
    >
      <Label
        for="lf_textfilterId"
        tag="label"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        Text_Filter
      </Label>
      <Input
        className="rounded-0"
        id="lf_textfilterId"
        name="lf_textfilter"
        onChange={[Function]}
        placeholder="Input Text For filtering..."
        type="text"
        value="bills"
      />
    </FormGroup>
  </Col>
  <Col
    md={4}
    sm={6}
    tag="div"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    <FormGroup
      tag="div"
    >
      <Label
        for="exampleSelectMulti"
        tag="label"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        Sorted_By
      </Label>
      <Input
        className="rounded-0 select-rightcorner"
        id="lf_sortbydateOramountId"
        name="lf_sortbydateOramount"
        onChange={[Function]}
        type="select"
        value="amount"
      >
        <option
          value="date"
        >
          Date
        </option>
        <option
          value="amount"
        >
          Amount
        </option>
      </Input>
    </FormGroup>
  </Col>
  <Col
    md={4}
    sm={6}
    tag="div"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    <Label
      tag="label"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      StartDate -- EndDate
    </Label>
    <withStyles(DateRangePicker)
      anchorDirection="left"
      appendToBody={false}
      block={false}
      calendarInfoPosition="bottom"
      customArrowIcon={null}
      customCloseIcon={null}
      customInputIcon={null}
      daySize={39}
      disableScroll={false}
      disabled={false}
      displayFormat={[Function]}
      enableOutsideDays={false}
      endDate={"1970-01-04T00:00:00.000Z"}
      endDateId="DpEndDate"
      endDatePlaceholderText="End Date"
      firstDayOfWeek={null}
      focusedInput={null}
      hideKeyboardShortcutsPanel={false}
      horizontalMargin={0}
      initialVisibleMonth={null}
      inputIconPosition="before"
      isDayBlocked={[Function]}
      isDayHighlighted={[Function]}
      isOutsideRange={[Function]}
      isRTL={false}
      keepFocusOnInput={false}
      keepOpenOnDateSelect={false}
      minimumNights={1}
      monthFormat="MMMM YYYY"
      navNext={null}
      navPrev={null}
      noBorder={false}
      numberOfMonths={1}
      onClose={[Function]}
      onDatesChange={[Function]}
      onFocusChange={[Function]}
      onNextMonthClick={[Function]}
      onPrevMonthClick={[Function]}
      openDirection="down"
      orientation="horizontal"
      phrases={
        Object {
          "calendarLabel": "Calendar",
          "chooseAvailableEndDate": [Function],
          "chooseAvailableStartDate": [Function],
          "clearDates": "Clear Dates",
          "closeDatePicker": "Close",
          "dateIsSelected": [Function],
          "dateIsUnavailable": [Function],
          "enterKey": "Enter key",
          "escape": "Escape key",
          "focusStartDate": "Interact with the calendar and add the check-in date for your trip.",
          "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
          "homeEnd": "Home and end keys",
          "jumpToNextMonth": "Move forward to switch to the next month.",
          "jumpToPrevMonth": "Move backward to switch to the previous month.",
          "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
          "keyboardShortcuts": "Keyboard Shortcuts",
          "leftArrowRightArrow": "Right and left arrow keys",
          "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
          "moveFocusByOneMonth": "Switch months.",
          "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
          "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
          "openThisPanel": "Open this panel.",
          "pageUpPageDown": "page up and page down keys",
          "questionMark": "Question mark",
          "returnFocusToInput": "Return to the date input field.",
          "selectFocusedDate": "Select the date in focus.",
          "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
          "upArrowDownArrow": "up and down arrow keys",
        }
      }
      readOnly={false}
      regular={false}
      renderCalendarInfo={null}
      renderDayContents={null}
      renderMonthElement={null}
      renderMonthText={null}
      reopenPickerOnClearDates={false}
      required={false}
      screenReaderInputMessage=""
      showClearDates={true}
      showDefaultInputIcon={false}
      small={false}
      startDate={"1970-01-01T00:00:00.000Z"}
      startDateId="DpStartDate"
      startDatePlaceholderText="Start Date"
      verticalHeight={null}
      verticalSpacing={22}
      weekDayFormat="dd"
      withFullScreenPortal={false}
      withPortal={false}
    />
    <FormText
      color="secondary"
      tag="small"
    >
      Select Date Range for expense.
    </FormText>
  </Col>
</div>
`;
